main.go

```diff
@@ -1,15 +1,52 @@
 package main
 
 import (
+	"context"
 	"flag"
 	"log"
 
+	"cloud.google.com/go/storage"
 	"github.com/fullstorydev/hauser/client"
 	"github.com/fullstorydev/hauser/config"
 	"github.com/fullstorydev/hauser/internal"
 	"github.com/fullstorydev/hauser/warehouse"
 )
 
+func MakeStorage(ctx context.Context, conf *config.Config) warehouse.Storage {
+	switch conf.Provider {
+	case config.LocalProvider:
+		return warehouse.NewLocalDisk(&conf.Local)
+	case config.AWSProvider:
+		return warehouse.NewS3Storage(&conf.S3)
+	case config.GCProvider:
+		gcsClient, err := storage.NewClient(ctx)
+		if err != nil {
+			log.Fatalf("Failed to create GCS client")
+		}
+		return warehouse.NewGCSStorage(&conf.GCS, gcsClient)
+	default:
+		log.Fatalf("unknown provider type: %s", conf.Provider)
+	}
+	return nil
+}
+
+func MakeDatabase(_ context.Context, conf *config.Config) warehouse.Database {
```

Comment:
I primarily went with this pattern so that I could more easily mock out each of these. I'll keep this in mind though as I continue to add functionality.

Reply:
I don't think using a first-class provider would make it harder to mock or test anything. I'm not suggesting that you make the provider an argument into any of the existing functions, so their tests wouldn't need to change. It would be an additional abstraction that can be used to keep warehouse-specific code out of `main.go`.

Reply:
Got it, I misunderstood what you meant. That's a good idea. I think there's a lot that can be done here to clean up this up, but I may follow up with that.